keps/83-workload-preemption/README.md

```diff
@@ -0,0 +1,666 @@
+- The resources from these Pods are not immediately available and they could
+  be arbitrarily delayed.
+- While Pods are terminating, a ClusterQueue's quota could be oversubscribed.
+
+The kubernetes Job status includes the number of Pending/Running Pods that are
+not terminating (don't have a `.metadata.deletionTimestamp`). We could use
+this information and write the old admission spec into an annotation to keep
+track of usage from non-terminating Pods. But this will be left for future work.
+
+### Increased admission latency
+
+Calculating and executing preemption is expensive. Potentially, every
+workload might benefit from preemption of running Workloads.
+
+To mitigate this, we will keep track of the minimum priority among the running
+Workloads in a ClusterQueue and cohort. If the minimum priorities are higher
+than or equal to the incoming Workload, we will skip preemption for it
+altogether.
+
+The assumption is that workloads with low priority are more common than
+workloads with higher priority and that Workloads are sent to ClusterQueues
+where most Workloads have the same priority.
+
+Additionally, the preemption algorithm is mostly a linear pass over the running
+workloads (plus sorting), so it doesn't add a significant complexity overhead
+over building the scheduling snapshot every cycle.
+
+The API updates from preemption will be executed in parallel.
+
+## Design Details
+
+The proposal consists of new API fields and a preemption algorithm.
+
+### ClusterQueue API changes
+
+The new API fields in ClusterQueue describe how to influence the selection
+of Workloads to preempt.
+
+```golang
+type ClusterQueueSpec struct {
+  ...
+  // preemption describes policies to preempt Workloads from ClusterQueues.
+  preemption ClusterQueuePreemption
+}
+
+type PreemptionByLowerPriorityWhenBorrowingPolicy string
+
+const (
+  PreemptionByLowerPriorityWhenBorrowingAllow = "Allow"
+  PreemptionByLowerPriorityWhenBorrowingDisallow = "Disallow"
+)
+
+type ClusterQueuePreemption struct {
+  // waitBeforePreemptionSeconds is the time in seconds that the ClusterQueue
+  // is willing to wait for Workloads to finish since the first time the
+  // Workload was attempted for admission, before preempting active Workloads
+  // with lower priority.
+  // The time when the Workload was first attempted for admission is present
+  // in the transitionTimestamp of the Admitted condition.
+  // If null, it means to wait indefinitely, effectively disabling preemption
+  // for incoming Workloads in this ClusterQueue.
+  WaitBeforePreemptionSeconds *int64
+
+  // byLowerPriorityWhenBorrowing determines whether pending
+  // Workloads in the cohort, that would fit in the min quota for their
+  // ClusterQueues, can preempt Workloads with higher priority in this
+  // ClusterQueue if it's using more than its min quota.
+  // Possible values are Allow and Disallow (default).
+  ByLowerPriorityWhenBorrowing *PreemptionByLowerPriorityWhenBorrowingPolicy
```

Comment:
Strange name, any precedence in k8s to this naming format?

Reply:
I don't think so.

Maybe something like this?

```yaml
preemption:
  onLowPriorityReclaim: Preempt|Ignore
```

Reply:
Aldo and I spent some time talking about the behavior and the name of this field.

Which of the following is this field controlling:
1. the policy _this_ queue will use to preempt workloads running in other queues to recover quota borrowed _from this queue_
2. the policy _other_ queues will use to preempt workloads running in this queue to recover quota this queue borrowed from _other queues_

Option 1 was what I expected, since different queues in a cohort could have different opinions about how important it is to recover quota borrowed from them, and option 2 can't express that.

If this is for option 1, it seems like there are at least three possible policies (having the behavior you get with a nil value for the policy be able to be represented as an explicit enum value is good):
* NeverPreempt
* PreemptLowerPriority
* PreemptImmediately

It's also unclear how this interacts with the `WaitBeforePreemptionSeconds` field. Can a queue be configured to reclaim borrowed quota by preemption (either immediately or by priority) without being forced to enable preemption among workloads in the queue?

Reply:
> Maybe something like this?
> 
> ```yaml
> preemption:
>   onLowPriorityReclaim: Preempt|Ignore
> ```

The "LowPriority" aspect seems like one of the possible policies, not something we should bake into the field name

Reply:
It is indeed option 1; I like the enum and we can perhaps name the field `preemptionPolicy` or just `policy`

Reply:
I changed it to:

```yaml
preepmtion:
  withinCohort: Never|FromLowerPriority|Always
  withinClusterQueue: Never|FromLowerPriority
```